### PR TITLE
refactor: use sdk.Dec for quantities

### PR DESCRIPTION
**Pull request by kingcre**
## Description

Use `sdk.Dec` instead of `sdk.Int`(also `sdk.DecCoin` instead of `sdk.Coin`) in x/exchange. This will reduce errors caused by decimal truncation.

## Tasks

- [x] Refactor codebase
- [x] Remove `QueueSendCoins` and `ExecuteSendCoins`, instead use new `Escrow` type
- [x] Write tests for decimal quantities

### CODE REVIEW

**Comment by crypin** [comment]
Please let me know when it ready for review

**Comment by kingcre** [comment]
@crypin It's now ready

File: x/amm/keeper/source.go [hunk 1]
```diff
@@ -157,46 +138,48 @@ func (k Keeper) AfterPoolOrdersExecuted(ctx sdk.Context, pool types.Pool, result
...
-				MulTruncate(types.DecMulFactor).
-				QuoTruncate(poolState.CurrentLiquidity.ToDec()))
+			feeGrowth := sdk.NewDecCoinFromDec(
+				fee.Denom, fee.Amount.ToDec().
+					MulTruncate(types.DecMulFactor).
+					QuoTruncate(poolState.CurrentLiquidity.ToDec()))
 			poolState.FeeGrowthGlobal = poolState.FeeGrowthGlobal.Add(feeGrowth)
 		} else if amtInDiff.IsNegative() { // sanity check
-			//panic(amtInDiff)
+			panic(amtInDiff)
```
Comment (crypin) [inline x/amm/keeper/source.go:171]: I think this part is likely to happen, I will share the detailed scenario again as soon as it is sorted out

**Review by crypin — CHANGES_REQUESTED** [review]

File: x/amm/keeper/invariants.go [hunk 2]
```diff
@@ -10,14 +10,19 @@ import (
 )
 
 func RegisterInvariants(ir sdk.InvariantRegistry, k Keeper) {
-	ir.RegisterRoute(types.ModuleName, "rewards-growth", RewardsGrowthInvariant(k))
+	ir.RegisterRoute(types.ModuleName, "rewards-growth-global", RewardsGrowthGlobalInvariant(k))
+	ir.RegisterRoute(types.ModuleName, "rewards-growth-outside", RewardsGrowthOutsideInvariant(k))
 	ir.RegisterRoute(types.ModuleName, "can-collect", CanCollectInvariant(k))
```
[the remark this answers is not in the document]
Comment (kingcre) [reply x/amm/keeper/invariants.go:15]: Actually we were already using cached context inside `CollectibleCoins`:
https://github.com/crescent-network/crescent/blob/e5a01f970a4c80ad9008a8413e6534e7b04be86a/x/amm/keeper/position.go#L187-L195

[the remark this answers is not in the document]
Comment (crypin) [reply x/amm/keeper/invariants.go:15]: Oh, I said it wrong, the state transition occurs without cached ctx in the `Collect` of this part, 

https://github.com/crescent-network/crescent/blob/e5a01f970a4c80ad9008a8413e6534e7b04be86a/x/amm/keeper/invariants.go#L118

so i think the determinism can be broken depending on whether or not there is Invariant checking config for each node

[the remark this answers is not in the document]
Comment (kingcre) [reply x/amm/keeper/invariants.go:15]: Oh, that's right. Fixed it.

**Comment by kingcre** [comment]
Merging this PR and the remaining issues will be handled in separate issues.

**Pull request merged**